Enhance stability calculation logic and refactor related code

leomanga opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #28010

#### Brief description of what is fixed or changed
This PR addresses the https://github.com/sympy/sympy/issues/28010 issue by introducing a new method based on Routh https://github.com/sympy/sympy/issues/28010#issuecomment-2849002351. Additionally, it implements the Extended Routh Table to prevent the appearance of infinitesimal elements during computation. It now also returns a list of auxiliary polynomials instead of just the first one found, and adds the `zero_col_infos` function to handle columns affected by the First Column Zero Case.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

leomanga committed:
physics.control: Refactor Routh-Hurwitz functions to use Poly consistently

leomanga committed:
physics.control: Remove unused import from routh_table.py

moorepants commented on sympy/physics/control/routh_table.py:
```diff
@@ -4,12 +4,11 @@
 __all__ = ['RouthHurwitz', 'neg_roots_conds']
 
-def neg_roots_conds(polynomial, var):
+def neg_roots_conds(polynomial: Poly):
```
Maybe this shouldn't be a public function if it requires a Poly. Poly objects are not used by the average user, but they are very likely to have an expression that happens to be a polynomial.

sylee957 replied:
There are other possible interfaces like accepting the list of coefficients, than Poly or expression. That can be more easier because that is native interface in Python. That makes sense because the result is independent of the symbol `s`

oscarbenjamin replied:
This should be a private function. The name should being with an underscore like `_neg_roots_conds`. It should also have a better name even if it is a private function.

leomanga replied:
I think it might be more convenient to make this function part of the public interface. That way, anyone who just wants to analyze a polynomial’s roots doesn’t have to go through the `TransferFunction` class if they don’t need its other capabilities. 

> It should also have a better name

You're right, I'll think of something better.

moorepants commented on sympy/physics/control/routh_table.py:
```diff
@@ -18,43 +17,36 @@ def neg_roots_conds(polynomial, var):
-    >>> neg_roots_conds(p3, s)
+    >>> p3 = Poly((s+1)*(s+a), s)
+    >>> neg_roots_conds(p3)
     [a + 1 > 0, False]
```
I prefer the prior example design if this is intended to be a public function.

leomanga commented:
I made the first minor changes. 

Here is the plan for the next changes:
 - [ ] Add inequalities with "<". The output could probably be `[[ineq with >], [ineq with <]]`. 
Symbols should correspond to either the first or the second inequality. Reading https://en.wikipedia.org/wiki/Routh%E2%80%93Hurwitz_stability_criterion#Higher-order_example, it might appear that `[ineq with >]` is sufficient, but they probably expect $a_n$ to be positive. In general the Routh-Hurwitz criterion counts the sign changes in the first column, and requiring them to be positive doesn't cover all possibilities.
 
- [ ] Find the best way to remove `cancel(num/den)` and create a list with only `num`, which will be used for the inequalities.

- [ ] If `zero_row_case == True`, force every symbolic expression in the same row to be != 0.

- [ ] Discuss the `limit` usage.

- [ ] Discuss the RouthHurwitz refactoring.

leomanga commented:
> I'm -1 to these changes. If we want these functions to be used by typical users of the controls package, then functions should accept expressions. It is fine if Poly() objects are used behind the scenes, but Poly() should not be input/output to public functions.

I agree with you.

leomanga committed:
physics.control: Extend stability conditions

leomanga commented:
In this PR I tried to solve the first point of the plan.
I'm not sure this approach is the best, but it should work.
The output of the function should be intended as `output[0] or output[1]`.

Another approach could simply be to simplify the leading term of the polynomial, in order to make the Routh-Hurwitz table begin with 1, which is positive and permits to use only inequalities using only ">". 
The problem of this approach is that the table will probably begin to generate very complicated expressions.

Let me know what you think.

moorepants commented:
>  The problem of this approach is that the table will probably begin to generate very complicated expressions.

This is prefered to it not working, not giving correct output, or being slow. The expressions being long is something the user can selectively try to deal with after use.

leomanga commented:
> This is prefered to it not working, not giving correct output, or being slow. The expressions being long is something the user can selectively try to deal with after use.

The approach used in my last PR should work, we can choose either. 
The leading term simplification approach returns better output, as produces only inequalities with ">".
The approach used in my last PR might result in simpler expressions.

Honestly I prefer the simplification approach, but I committed the other one to illustrate the alternatives I found.

leomanga committed:
Refactor Routh-Hurwitz conditions function and update related tests

- Renamed `neg_roots_conds` to `negative_real_root_conditions` for clarity.
- Updated references in `lti.py` and `routh_table.py` to use the new function name.
- Enhanced the `negative_real_root_conditions` function to handle zero column cases.

leomanga committed:
physics.control: Implement the Extended Routh's Table

sylee957 commented on sympy/physics/control/routh_table.py:
```diff
@@ -1,93 +1,94 @@
-    >>> neg_roots_conds(p3, s)
-    [a + 1 > 0, False]
+    >>> negative_real_root_conditions(p3, s)
+    [[a + 1 > 0, False], [False]]
```
Can you add document about how to interpret this result?
I think that the result had changed and I'm not sure about the meaning of this array by 
`Returns the conditions for a polynomial to have all its roots with negative real parts.`

leomanga replied:
https://github.com/sympy/sympy/pull/28021#issuecomment-2847565623

>The output of the function should be intended as output[0] or output[1].

I haven't written the documentation yet because I didn't know if this approach was the best (https://github.com/sympy/sympy/pull/28021#issuecomment-2847687694).

However, I expect to return to the previous output form after the alternative Routh-Hurwitz test implementation you proposed.

leomanga commented:
I started implementing the new approach with this diff

```diff
-def negative_real_root_conditions(polynomial, var):
+def negative_real_root_conditions(
+        polynomial: Expr | Poly,
+        var: Symbol | None = None, /,
+        *, domain=None, sqf: bool = False) -> list[StrictGreaterThan | bool]:
+    p: Poly = Poly(polynomial, var, domain=domain)
+
+    _, p = p.clear_denoms(convert=True)
+
+    coeffs = p.rep.to_list()
+    new_domain = p.domain
+
+    return _stable(coeffs, new_domain, sqf=sqf)
+
+def _stable(p, domain, *, sqf) -> list[StrictGreaterThan | bool]:
+    if len(p) < 2:
+        return [True]
+
+    if len(p) == 2:
+        p0, p1 = p
+        return [domain.to_sympy(p0 * p1) > 0]
+
+    p0, p1, *_ = p
+
+    cond = p0 * p1
+
+    qs = [p1 * qi for qi in p]
+    for i in range(1, len(p), 2):
+        qs[i - 1] -= p[i] * p0
+    qs = qs[1:]
+
+    if sqf:
+        qs = dup_sqf_part(qs, domain)
+
+    return [domain.to_sympy(cond) > 0]+ _stable(qs, domain, sqf=sqf)
```

but I found this problem:

```python
In [0]: p = (s**2 + 1)**2

In [3]: negative_real_root_conditions(p, s)
Out[3]: [False, False, False, False]

In [4]: negative_real_root_conditions(p, s, sqf=True)
ZeroDivisionError
```

It happens because, during the process, `qs` become a list of zeros.

I tried to fix it with
```python
    if domain.is_nonpositive(cond):
        return [False]
```
With these lines, the program should also be faster, terminating even when it finds a negative condition.
The problem here is in this example
```python
In [1]: s = symbols("s")

In [2]: b1, b2, b3, b4 = symbols("b1:5")

In [3]: p = b1*s**3 + b2*s**2 + b3*s + b4

In [4]: negative_real_root_conditions(p, s)
Out[4]: [b1*b2 > 0, False]

```

Where I expect 
```python
Out[8]: 
⎡                  2        3             3      4       2   2      3          3   2   2     4   3       ⎤
⎣b₁⋅b₂ > 0, - b₁⋅b₂ ⋅b₄ + b₂ ⋅b₃ > 0, - b₁ ⋅b₂⋅b₄  + 3⋅b₁ ⋅b₂ ⋅b₃⋅b₄  - 3⋅b₁⋅b₂ ⋅b₃ ⋅b₄  + b₂ ⋅b₃ ⋅b₄ > 0⎦
```

The problem is in `dom.is_nonpositive()` (I tried to replicate the second step of the function):
```python
In [1]: s = symbols("s")

In [2]: b1, b2, b3, b4 = symbols("b1:5")

In [3]: p = Poly(-b1*b2**2*b4 + b2**3*b3, s)

In [4]: coeffs = p.rep.to_list()

In [5]: dom = p.domain

In [6]: coeffs
Out[6]: [-b1*b2**2*b4 + b2**3*b3]

In [7]: dom.is_nonpositive(coeffs[0])
Out[7]: True

```
Why is the output `True`? Is it normal?
`coeffs[0]` is a `PolyElement`, maybe I didn't understand something about it.

leomanga commented:
I tried to compute the inequalities for the charpoly of the matrix `A` used here https://github.com/sympy/sympy/issues/28010#issue-3029999340, but the program hangs.
Printing the various conditions during the recursion I can see that the expression complexity grows very fast.

This is the code I used:
```python
def negative_real_root_conditions(
        polynomial: Expr | Poly,
        var: Symbol | None = None, /,
        *, domain=None, sqf: bool = False) -> list[StrictGreaterThan | bool]:
    p: Poly = Poly(polynomial, var, domain=domain)

    _, p = p.clear_denoms(convert=True)

    coeffs = p.rep.to_list()
    new_domain = p.domain

    return _stable(coeffs, new_domain, sqf=sqf)

def _stable(p, domain, *, sqf) -> list[StrictGreaterThan | bool]:
    if len(p) < 2:
        return [True]

    if len(p) == 2:
        p0, p1 = p
        return [domain.to_sympy(p0 * p1) > 0]

    p0, p1, *_ = p

    cond = p0 * p1
    if cond == 0:
        return [False]

    qs = [p1 * qi for qi in p]
    for i in range(1, len(p), 2):
        qs[i - 1] -= p[i] * p0
    qs = qs[1:]

    if sqf:
        qs = dup_sqf_part(qs, domain)

    return [domain.to_sympy(cond) > 0] + _stable(qs, domain, sqf=sqf)
```

The old implementation seems to work fine without the `cancel` call in the `RouthHurwtz` class and with other examples.

Maybe we could use the old algorithm for the `lti` methods as default, and add this new algorithm for free-division conditions.

Let me know if it seems reasonable.

oscarbenjamin commented:
A more detailed comparison of different approaches and different example systems is needed.

So far in the discussion above it is not clear to me whether any of the implementations being compared is even correct which is the first thing that needs to be established before considering speed.

If the "old implementation" is the version that tries to check whether things are positive or negative then it is never going to work in the symbolic case.

It is likely that this should have separate implementations for the purely numeric case vs the symbolic case rather than trying to make a single implementation that handles both.

leomanga commented:
I tried making some checks and changes and I found some problems and possible solutions.

The first thing is that, for some specific cases, the implementation with the Routh-Hurwitz table doesn't work. This is related to the "first zero column" case. The only way to solve that without the infinitesimal element is very tricky and I don't think it's worth trying to implement it.

The second thing is that I tried to write the "new implementation" using only expressions and everything seems to work. 
Is there any issue with using only expressions instead of working with domains?
I still have a lot to learn about sympy, so I'm not really sure about the answer.

This is the code I tried (I removed the ">" in the conditions to show numbers)

```python
def negative_real_root_conditions(
        polynomial: Expr | Poly,
        var: Symbol | None = None, /,
        *, domain=None, force_no_div = False) -> list[StrictGreaterThan | bool]:
    p: Poly = Poly(polynomial, var, domain=domain)

    _, p = p.clear_denoms(convert=True)

    coeffs = p.all_coeffs()

    if all(c.is_number for c in coeffs) and force_no_div is False:
        return _calc_conditions_div(coeffs)

    return _calc_conditions_no_div(coeffs)

def _run_checks(func):
    """
    Decorator to provide the basic checks for the recursive
    stability functions.

    """
    def wrapper(p: list) -> list[StrictGreaterThan | bool]:
        if len(p) < 2:
            return []

        if (p[0]*p[1]).is_nonpositive:
            return [-1]

        if len(p) == 2:
            return [p[0] * p[1]]

        return func(p)

    return wrapper

@_run_checks
def _calc_conditions_div(p: list) -> list[StrictGreaterThan | bool]:
    """
    Stability check with divisions, used for numeric polynomials.
    Algorithm from:
    https://courses.washington.edu/mengr471/resources/Routh_Hurwitz_Proof.pdf

    """
    qs = p.copy()
    for i in range(1, len(p), 2):
        qs[i - 1] -= p[i] * p[0] / p[1]
    qs = qs[1:]

    return [p[0] * p[1]] + _calc_conditions_div(qs)

@_run_checks
def _calc_conditions_no_div(p: list) -> list[StrictGreaterThan | bool]:
    """
    Stability check without divisions, used for symbolic polynomials.
    Algorithm is a rivisitation of the one in:
    https://courses.washington.edu/mengr471/resources/Routh_Hurwitz_Proof.pdf

    """
    qs = [p[1] * qi for qi in p]
    for i in range(1, len(p), 2):
        qs[i - 1] -= p[i] * p[0]
    qs = qs[1:]

    return [p[0] * p[1]] + _calc_conditions_no_div(qs)
```
I implemented both `_calc_conditions_no_div` and `_calc_conditions_div` for purely numeric polynomials, to avoid calculating exponentially growing numbers.
```python
In [1]: negative_real_root_conditions((s+1)*(s+2)*(s+13)*(s+11.2)*(s+3), s)
Out[1]: [30.2, 7968.56, 250450.572715232, 1400064.85198831, 1288580.79185398]

In [2]: negative_real_root_conditions((s+1)*(s+2)*(s+13)*(s+11.2)*(s+3), s, force_no_div = True)
Out[2]: [30.2, 7267645.4624, 1.45042610999498e+16, 4.2305181546161e+33, 2.55976755070878e+67]

```
Obviously, things discussed here https://github.com/sympy/sympy/issues/28010#issuecomment-2849225601 remains. It  should be possible to simplify expressions. This, assuming that implementation works, could be done at a later stage and, for now, we could leave the problem of simplifying expressions to the user.

leomanga committed:
physics.control: Refactor stability condition functions and update related tests

leomanga committed:
physics.control: Remove unused imports from routh_table.py and test_routh_table.py

leomanga committed:
physics.control: Remove unused import 'Or' from lti.py

leomanga committed:
physics.control: Remove '|' operator for type hints for Python 3.9 compatibility

leomanga committed:
physics.control: Update references in RouthHurwitz class docstring

leomanga committed:
physics.control: Remove '|' operator for type hints for Python 3.9 compatibility

leomanga committed:
physics.control: Fix formatting in docstring for negative_real_root_conditions

leomanga committed:
physics.control: Remove return type hints from stability check functions

leomanga commented:
> Is that what you are doing?
That's what I did in most cases for the previous algorithm.
Now I've just changed the list of inequalities to match the new output. The important thing is to return the same result as before, which seems to be the case. It can be verified easily when `reduce_inequalities` is called and through the `is_stable` tests and doctests.

moorepants commented on sympy/physics/control/tests/test_routh_table.py:
```diff
@@ -11,8 +12,7 @@ def test_table():
     # generic polynomial tests
-    epsilon = symbols('epsilon')
-    t1 = RouthHurwitz(p1, s, epsilon)
+    t1 = RouthHurwitz(p1, s)
```
Why is epsilon no longer needed/used?

moorepants commented on sympy/physics/control/routh_table.py:
```diff
@@ -216,34 +241,32 @@ class RouthHurwitz(MutableDenseMatrix):
     True
-    >>> RouthHurwitz(p2, s).auxiliary_polynomial
-    Poly(2*s**4 + 12*s**2 + 16, s, domain='ZZ')
+    >>> RouthHurwitz(p2, s).auxiliary_polynomials
```
Is this plural because it can now return more than one?

leomanga replied:
Previously, it returned only the first auxiliary polynomial found, even though there can be multiple ones.

leomanga commented on sympy/physics/control/routh_table.py:
```diff
@@ -1,93 +1,121 @@
-def neg_roots_conds(polynomial, var):
+def negative_real_root_conditions(
+        polynomial,
+        var = None, /,
```
It's used by @oscarbenjamin https://github.com/sympy/sympy/issues/28010#issuecomment-284922560.
It means that the parameters before the slash must be passed positionally.

sylee957 replied:
I think that distinguishing positional-only arguments is sometimes useful though, especially for overriding classes.
For example: you should learn that this is not compatible override
```
class Foo:
   def foo(self, x):
      ...

class Bar(Foo):
    def foo(self, y):
      ...
```
because base method allows to use like `.foo(x=...)` and it won't work with `Bar` class.
It is also confusing that Python language has features like calling non-optional parameters by name, as well as optional parameters can be positionally called. That some people find that as problem than convenience. 
(and that's why probably the PEP had been suggested).

I do think that we should use that more in SymPy when we encounter non-semantical parameters like `x, y` or `a, b`.
But for parameters named like `polynomial`, the name contains semantics and I do agree that we don't need it for this case.

moorepants replied:
If this is a SymPy design guideline, it should be in the docs somewhere.

sylee957 replied:
I think you should rather suggest a style guide if you want to argue against using `/` just because you don't find it pretty or are not familiar with it.

Anyway, I'd appreciate seeing code written in more robust syntax, especially for people who are new to Python or SymPy.
On the other hand, rejecting the idea just because it's unfamiliar to you is not very helpful. If you stick to that kind of thinking, fewer people will want to work together with you.

leomanga commented on sympy/physics/control/routh_table.py:
```diff
@@ -1,93 +1,121 @@
+def negative_real_root_conditions(
+        polynomial,
+        var = None, /,
+        *, domain=None):
```
No, they are not needed. If you think they complicate things, I can remove them.

moorepants commented:
Also, please open a new issue about the performance of the Routh table for large matrices or matrices with large expressions. This can be solved in a separate PR. This should be able to work with large expressions, with little performance degradation.

leomanga committed:
physics.control: Fix formatting and add negative_real_root_conditions in the docs

oscarbenjamin commented:
For a quadratic polynomial with coefficients `a`, `b` and `c` the condition is that `a`, `b` and `c` must have the same sign. Here we get:
```
In [1]: from sympy.physics.control import negative_real_root_conditions

In [2]: a, b, c = symbols('a, b, c')

In [3]: p = a*x**2 + b*x + c

In [4]: negative_real_root_conditions(p, x)
Out[4]: 
⎡          3      ⎤
⎣a⋅b > 0, b ⋅c > 0⎦
```
Obviously the `b**3` can just be `b`.

For a cubic it is:
```
In [5]: a, b, c, d = symbols('a, b, c, d')

In [6]: p = a*x**3 + b*x**2 + c*x + d

In [7]: negative_real_root_conditions(p, x)
Out[7]: 
⎡          2                                   3    ⎤
⎣a⋅b > 0, b ⋅(-a⋅d + b⋅c) > 0, b⋅d⋅(-a⋅d + b⋅c)  > 0⎦
```
That is `a` and `b` have to have the same sign. The `b**2` is not needed so we have `b*c - a*d > 0`. Then that factor is not needed in the final inequality which can become `b*d > 0`. So `a`, `b` and `d` need to have the same sign (take positive) and we need `c > a*d/b`.

As you build up to larger polynomials it looks like more and more redundant factors are in the inequalities:
```python
In [19]: a, b, c, d, e, f = symbols('a, b, c, d, e, f')

In [20]: p = Poly([a, b, c, d, e, f], x).as_expr()

In [21]: for c in negative_real_root_conditions(p, x): pprint(c)
a⋅b > 0
 2                 
b ⋅(-a⋅d + b⋅c) > 0
            2 ⎛   2                                ⎞    
(-a⋅d + b⋅c) ⋅⎝- b ⋅(-a⋅f + b⋅e) + b⋅d⋅(-a⋅d + b⋅c)⎠ > 0
                                      2                                                                                             
⎛   2                                ⎞  ⎛                  3                             ⎛   2                                ⎞⎞    
⎝- b ⋅(-a⋅f + b⋅e) + b⋅d⋅(-a⋅d + b⋅c)⎠ ⋅⎝- b⋅f⋅(-a⋅d + b⋅c)  + (-a⋅d + b⋅c)⋅(-a⋅f + b⋅e)⋅⎝- b ⋅(-a⋅f + b⋅e) + b⋅d⋅(-a⋅d + b⋅c)⎠⎠ > 0
                                                                                                                                                3    
                 ⎛   2                                ⎞ ⎛                  3                             ⎛   2                                ⎞⎞     
b⋅f⋅(-a⋅d + b⋅c)⋅⎝- b ⋅(-a⋅f + b⋅e) + b⋅d⋅(-a⋅d + b⋅c)⎠⋅⎝- b⋅f⋅(-a⋅d + b⋅c)  + (-a⋅d + b⋅c)⋅(-a⋅f + b⋅e)⋅⎝- b ⋅(-a⋅f + b⋅e) + b⋅d⋅(-a⋅d + b⋅c)⎠⎠  > 0
```
Squared factors can be removed. Cube factors don't need to be cubed. 

Things that are already known positive could be removed. This last point is significant because if it can be assumed that `a` is positive then everything simplifies more:
```python
In [33]: a, b, c, d, e, f = symbols('a, b, c, d, e, f')

In [34]: p = Poly([1, b, c, d, e, f], x).as_expr()

In [35]: for c in negative_real_root_conditions(p, x): pprint(c)
b > 0
 2              
b ⋅(b⋅c - d) > 0
         2 ⎛   2                          ⎞    
(b⋅c - d) ⋅⎝- b ⋅(b⋅e - f) + b⋅d⋅(b⋅c - d)⎠ > 0
                                2                                                                              
⎛   2                          ⎞  ⎛               3                       ⎛   2                          ⎞⎞    
⎝- b ⋅(b⋅e - f) + b⋅d⋅(b⋅c - d)⎠ ⋅⎝- b⋅f⋅(b⋅c - d)  + (b⋅c - d)⋅(b⋅e - f)⋅⎝- b ⋅(b⋅e - f) + b⋅d⋅(b⋅c - d)⎠⎠ > 0
                                                                                                                        3    
              ⎛   2                          ⎞ ⎛               3                       ⎛   2                          ⎞⎞     
b⋅f⋅(b⋅c - d)⋅⎝- b ⋅(b⋅e - f) + b⋅d⋅(b⋅c - d)⎠⋅⎝- b⋅f⋅(b⋅c - d)  + (b⋅c - d)⋅(b⋅e - f)⋅⎝- b ⋅(b⋅e - f) + b⋅d⋅(b⋅c - d)⎠⎠  > 0
```
Off the bat we see that `b` is positive which means any power of `b` can be factored out from every lhs. Then we immediately see that `(b*c - d)` is positive so any power of it can be divided away. Then we get a bigger expression from which a factor of `b` could be removed. The final inequality looks complicated but when you see how many factors are already known to be positive it simplifies to just `f > 0`.

It looks like this method causes the expressions to grow but it would be possible to control that growth if working with assumed positive expressions (starting from the leading coefficient `a`):
```python
In [78]: cs = negative_real_root_conditions(p, x)

In [79]: cs = [e.lhs for e in cs]

In [80]: cs[1] = cs[1] / cs[0]**2

In [81]: cs[2] = factor_terms(cs[2] / cs[1]**2 / cs[0])

In [82]: cs[3] = factor_terms(cs[3]) / cs[0]**3 / cs[1] / cs[2]**2

In [83]: cs[4] = factor_terms(cs[4]) / cs[0]**5 / cs[1]**4 / cs[2] / cs[3]**3

In [84]: for cond in cs: print(cond > 0)
b > 0
b*c - d > 0
-b*(b*e - f) + d*(b*c - d) > 0
-f*(b*c - d)**2 + (b*e - f)*(-b*(b*e - f) + d*(b*c - d)) > 0
f > 0
```
I think it is also possible to further rearrange these into some standard-ish form. This form also makes it look like the minors of 

It possibly makes sense for the method to control that growth as it goes along though rather than just at the end. Here we just have symbols for the coefficients but with large symbolic expressions there it would be difficult to understand what is happening.

leomanga committed:
physics.control: Return sympy Boolean object instead of python bool

leomanga committed:
physics.control: var, in negative_real_root_conditions, is no longer optional

moorepants commented:
> Do I have to add the calculation here https://github.com/sympy/sympy_benchmarks/ ?

No, just add a test here.

moorepants commented:
I would like to for us to get this fixed merge so that #28010 is addressed and encourage other changes, like moving core routines to the polys module (or other sympy core areas), making the results simpler, design changes, or addressing edge cases for other issues/prs. Please keep in mind that the goal is to have a basic way to provide stability conditions for the control package that, at least initially, works for typical problems control engineers might face. The PR is already 2 months old and the GSoC period is only so long.

leomanga commented:
Thanks @oscarbenjamin for your explanation, things are clearer now.
I had just assumed that `is_nonpositive` does exactly what its name suggests. Maybe it would be useful to mention this behavior in the documentation.
Anyway, I agree with @moorepants, this could probably be addressed in a separate PR.
I agree that the current approach could be improved, but even if the check with `is_nonpositive` fails, the expression will still appear in the list of conditions multiplied by another one. So we don’t lose any information, the output will still be ideally correct. We'll just have a `False` condition, which correctly means that the system is not asymptotically stable.

leomanga committed:
physics.control: Add test for issue #28010

leomanga committed:
Add newline for quality

oscarbenjamin commented:
> You can also add a benchmark for this, but it will only be run when the benchmarks are run (I'm not sure if they run in this CI anymore, as I haven't seen the PR comments in a long time).

They are still run, just the job that comments on the PR was disabled because it was insecure. You can see the outputs if you click on the benchmarks job e.g. for a recent commit here:
https://github.com/sympy/sympy/actions/runs/15794256170/job/44524573388

leomanga committed:
physics.control: Rename negative_real_root_conditions to negative_real_part_conditions

oscarbenjamin commented:
I believe that the methods here are mathematically correct although the theorem they are based on requires a precondition that the leading coefficient is not zero that is not checked/returned here:
```python
In [30]: zero = sin(1)**2 + cos(1)**2 - 1

In [31]: And(*negative_real_part_conditions(zero*(x+1), x)).simplify()
Out[31]: True

In [32]: And(*negative_real_part_conditions(zero*(x-1), x)).simplify()
Out[32]: False
```
It could be stated simply as a premise of the method though that the leading coefficient is assumed nonzero. In the case of the characteristic polynomial of a matrix the leading coefficient is always 1 so we don't have a problem with this.

It does not quite make sense to just return `Ne(LC, 0)` as part of the conditions because if it is zero then the roots of the lower degree polynomial may still be in the left half plane. Typically though any case where the leading coefficient is zero would be a degenerate case in some other way where it is not straight-forward to analyse based on the remaining roots so the premise of nonzero leading coefficient is implied for most analysis.

I don't think there is any benefit in using the division based method for any possible inputs so the division free method could always be used. I suspect that would also be more accurate for float inputs since division is an important source of avoidable rounding error.

leomanga commented:
>I don't think there is any benefit in using the division based method for any possible inputs so the division free method could always be used. I suspect that would also be more accurate for float inputs since division is an important source of avoidable rounding error.

I'm not sure it's a good idea to build these numbers by default.
```python
In [4]: negative_real_part_conditions((s+1)*(s+2)*(s+13)*(s+11.2)*(s+3)*(s+3.4)*(s+1.2), s)
Out[4]: 
[34.8, 15527644.85376, 1.40475220535259e+17, 1.51600390593604e+36, 3.03678542582823e+73, 2.26240636716244e+147, 1.88377077566808e+294]
```
Since we are in a physics module, it might be acceptable to treat floating-point numbers that are very close to zero as zero. Physically, the system's behavior will be essentially the same.
However, we could consider providing an option or an alternative division free method just for numbers. (I'm more inclined toward the second option, as it makes it clear that the implementation is specifically designed for numeric use cases).

leomanga committed:
physics.control: Update docstrings to clarify assumptions about leading coefficients

leomanga committed:
Add newline in docstring

leomanga commented:
Since the method doesn’t seem to have any algorithmic errors, can we merge this PR?
I will open these issues to address the various problems we discussed. Let me know if I missed anything:
- Extend and adapt `negative_real_part_conditions` to work with generic polynomial domains.
- Refactor the `RouthHurwitz` class to use composition with the Matrix class.

@moorepants @oscarbenjamin @sylee957

sylee957 commented:
I'm okay with the changes though. I agree that extended routh table is at least better for symbolic computation because we don't need to worry about extra expression blowups with `epsilon`.

oscarbenjamin commented:
The only thing I object to before merging is the 200 line polynomial. It is written in an absurdly long way.

Things to follow up:

1. The implementation while giving correct results has massive expression growth.
2. The RouthHuwitz class is not used any more and I think it still returns incorrect results. Is there a reason not to just remove the whole class? (Perhaps it could just be a function that returns a Matrix and another function that checks stability but only for a matrix of rational numbers.)

leomanga committed:
physics.control: Refactor complex polynomial test

leomanga committed:
physics.control: Remove unused imports from test_routh_table.py

leomanga commented:
> The only thing I object to before merging is the 200 line polynomial. It is written in an absurdly long way.

Solved in my last PR

moorepants commented:
Ok, I'm merging based on the above comments. @leomanga has opened issues for the outstanding concerns.
